File: backend/src/main/java/com/novelwriting/entity/Foreshadow.java
package com.novelwriting.entity;

import jakarta.persistence.*;
import lombok.Data;
import java.time.LocalDateTime;
import java.util.HashSet;
import java.util.Set;

@Data
@Entity
@Table(name = "foreshadows")
public class Foreshadow {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(name = "novel_id")
    private Long novelId;

    @Column(nullable = false, length = 200)
    private String title;

    @Column(columnDefinition = "TEXT")
    private String content;

    @Column(name = "laid_at", length = 100)
    private String laidAt;

    @Column(name = "revealed_at", length = 100)
    private String revealedAt;

    @Column(length = 20)
    private String status = "pending";

    @ManyToMany
    @JoinTable(
        name = "foreshadow_tags",
        joinColumns = @JoinColumn(name = "foreshadow_id"),
        inverseJoinColumns = @JoinColumn(name = "tag_id")
    )
    private Set<Tag> tags = new HashSet<>();

    @Column(name = "created_at", updatable = false)
    private LocalDateTime createdAt;

    @Column(name = "updated_at")
    private LocalDateTime updatedAt;

    @PrePersist
    protected void onCreate() {
        createdAt = LocalDateTime.now();
        updatedAt = LocalDateTime.now();
    }

    @PreUpdate
    protected void onUpdate() {
        updatedAt = LocalDateTime.now();
    }
}

File: frontend-vue/src/types/foreshadow.ts
import type { ID, Timestamp } from './index'
import type { Tag } from './tag'

export type ForeshadowStatus = 'pending' | 'revealed' | 'abandoned'

export interface Foreshadow {
  id: ID
  title: string
  content?: string
  status: ForeshadowStatus
  laidAt?: string
  revealedAt?: string
  tags?: Tag[]
  createdAt?: Timestamp
  updatedAt?: Timestamp
}

export interface ForeshadowCreateDTO {
  title: string
  content?: string
  status?: ForeshadowStatus
  laidAt?: string
  revealedAt?: string
}

export interface ForeshadowUpdateDTO extends Partial<ForeshadowCreateDTO> {}
